### PR TITLE
Adding several serial functions and unit tests for all print and println functions

### PR DESCRIPTION
**Pull request by frznlogic**
Added mocking for:
* Serial::read,  
* Serial::available, 
* Serial::write
* A new helper class stringCapture which simplifies getting all data from Serial functions, or other functions for that matter. This helps when testing code with arbitrary number of calls to Serial functions, but where the "sent" string is known (ie, you know how the protocol should look, but don't care how it is created). 

Unit tests are added for almost all functionality, and also updated to cover all print/println functions.

### CODE REVIEW

File: include/arduino-mock/serial.h [hunk 1]
```diff
@@ -91,4 +100,26 @@ class Serial_ {
...
 void releaseSerialMock();
 
+
+/*
+ * Thsi is a helper class which can be Invoked on EXPECT_CALL to gather data
+ * from repeated calls to Serial functions. For example usage, see the unit
+ * tests in test/serial_unittest.cc
+ *
+ */
+class stringCapture {
```
Comment (ikeyasu) [inline include/arduino-mock/serial.h]: Why do we need `stringCapture`?
I think ... it is not mock for Arduino.

**Comment by ikeyasu** [comment]
Good work!
My comment is only one for a position where we should have `stringCapture`.
It seems a helper function for `arduino-mock`'s unittest, so it should be locate in /test?

**Comment by frznlogic** [comment]
Hi Ikeyasu! 

Thanks! I really appreciate the project fyi, I've been looking for a good mocking system of the arduino platform for a long time as it is fairly hard to test on the platform itself :). 

Anyways, back to topic, the stringCapture class is supposed to be used by external users. I'm currently using it in a private project to verify functions which calls Serial::write arbitrary number of times and where I don't want to rely on the internal implementation of the function, but the outcoming stream format is really important. In this testcase, I encode it using the arduino functions, then I decode it using the PC side decoder within the unit test/mock framework.

File: include/arduino-mock/serial.h (hunk 1 above)
Comment (frznlogic) [inline include/arduino-mock/serial.h]: Hi Ikeasu, I wrote a longer response below. The stringCapture stuff is supposed to be used by external testers. The naming is a bit bad, any thoughts on a new name?

**Comment by ikeyasu** [comment]
Thank you for your reply.
I would like to improve Arduino testing framework. (You know the current Arduino testing is very poor..)

>  I'm currently using it in a private project to verify functions which calls Serial::write arbitrary number of times and where I don't want to rely on the internal implementation of the function

You mean you are using it as a library for your Arduino project?
If so, please consider to separate the class to another header file.
I understood that It is necessary but I would like to keep a structure of 'serial.h' as same as Arduino's library.

**Comment by frznlogic** [comment]
Fair enough. I've moved the functions into a file called serialHelper.cc/h and serialHelper_unittest.cc. Makefiles have been updated. Is this better?

**Comment by ikeyasu** [comment]
LGTM. It was merged manually.
Thanks a lot.

**Comment by frznlogic** [comment]
Thank you :)

**Pull request merged**